### PR TITLE
Issue-232: Making it possible to reset a single scope

### PR DESCRIPTION
**Pull request by dlemures**
Issue: https://github.com/stephanenicolas/toothpick/issues/232

### CODE REVIEW

File: toothpick-runtime/src/main/java/toothpick/ScopeNode.java [hunk 1]
```diff
@@ -200,6 +197,14 @@ public boolean isBoundToScopeAnnotation(Class<? extends Annotation> scopeAnnotat
     return scopeAnnotationClasses.contains(scopeAnnotationClass);
   }
 
+  /**
+   * Resets the state of the scope
```
Comment (stephanenicolas) [inline toothpick-runtime/src/main/java/toothpick/ScopeNode.java]: missing period at the end of javadoc. You will burn in hell. 
Add this is for testing only.

File: toothpick-runtime/src/main/java/toothpick/ScopeImpl.java [hunk 2]
```diff
@@ -522,10 +522,18 @@ private void crashIfClosed() {
     }
   }
 
-  static void reset() {
+  static void resetUnBoundProviders() {
     mapClassesToUnNamedUnBoundProviders.clear();
   }
 
+  @Override
+  protected void reset() {
```
Comment (stephanenicolas) [inline toothpick-runtime/src/main/java/toothpick/ScopeImpl.java]: I would copy the javadoc here too.

File: toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java [hunk 3]
```diff
@@ -217,6 +217,10 @@ public void providerGet_shouldFail_whenScopeIsClosed_andThereAreNoDependencies()
     //THEN
   }
 
+  @Test
+  public void reset_should() throws Exception {
+  }
```
Comment (stephanenicolas) [inline toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java]: empty test

File: toothpick-runtime/src/test/java/toothpick/ScopeNodeTest.java [hunk 4]
```diff
@@ -423,4 +424,19 @@ public void testGetParentScopeNames_shouldReturnParentNames_whenThereAreNoParent
...
+    scope.close();
+
+    //WHEN
+    scope.reset();
+
+    //THEN
+    assertThat(scope.isBoundToScopeAnnotation(CustomScope.class), is(false));
+    assertThat(scope.isOpen, is(true));
+  }
 }
```
Comment (stephanenicolas) [inline toothpick-runtime/src/test/java/toothpick/ScopeNodeTest.java]: end of line missing

Comment (dlemures) [inline toothpick-runtime/src/test/java/toothpick/ScopeNodeTest.java]: with love

File: toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java (hunk 3 above)
Comment (dlemures) [inline toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java]: with love

Comment (stephanenicolas) [inline toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java]: :)

File: toothpick-runtime/src/main/java/toothpick/Toothpick.java [hunk 5]
```diff
@@ -126,7 +127,17 @@ public static void reset() {
...
     }
     ConfigurationHolder.configuration.onScopeForestReset();
-    ScopeImpl.reset();
+    ScopeImpl.resetUnBoundProviders();
+  }
+
+  /**
+   * Reset the state of a single scope. Useful for automation testing when we want to reset the scope used to install
+   * test modules.
+   * @param scope we want to reset
```
Comment (stephanenicolas) [inline toothpick-runtime/src/main/java/toothpick/Toothpick.java]: the scope we want to reset.

Comment (stephanenicolas) [inline toothpick-runtime/src/main/java/toothpick/Toothpick.java]: PERIOD

File: toothpick-runtime/src/test/java/toothpick/ScopeNodeTest.java [hunk 6]
```diff
@@ -423,4 +424,19 @@ public void testGetParentScopeNames_shouldReturnParentNames_whenThereAreNoParent
...
+    scope.close();
+
+    //WHEN
+    scope.reset();
+
+    //THEN
+    assertThat(scope.isBoundToScopeAnnotation(CustomScope.class), is(false));
+    assertThat(scope.isOpen, is(true));
+  }
+}
```
Comment (stephanenicolas) [inline toothpick-runtime/src/test/java/toothpick/ScopeNodeTest.java]: EOL

File: toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java [hunk 7]
```diff
@@ -217,6 +220,23 @@ public void providerGet_shouldFail_whenScopeIsClosed_andThereAreNoDependencies()
     //THEN
   }
 
+  @Test(expected = NoFactoryFoundException.class)
+  public void reset_shouldResetBoundProviders_andFlagTheTestModuleToFalse() throws Exception {
+    //GIVEN
+    ScopeImpl scope = new ScopeImpl("root");
+    scope.installTestModules(new Module() { {
+      bind(IFoo.class).to(Foo.class);
+      bind(Bar.class).withName("name").to(BarChild.class);
```
Comment (stephanenicolas) [inline toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java]: not used

File: toothpick-runtime/src/test/java/toothpick/ScopeNodeTest.java (hunk 6 above)
Comment (dlemures) [inline toothpick-runtime/src/test/java/toothpick/ScopeNodeTest.java]: we dont follow that pattern on other test classes i saw

File: toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java (hunk 7 above)
Comment (dlemures) [inline toothpick-runtime/src/test/java/toothpick/ScopeImplTest.java]: ohh i didn't use it in the end 😅

**Pull request merged**